crates/typst-pdf/src/lib.rs

```diff
@@ -274,6 +288,45 @@ fn write_catalog(ctx: &mut PdfContext, ident: Option<&str>, timestamp: Option<Da
     if let Some(lang) = lang {
         catalog.lang(TextStr(lang.as_str()));
     }
+
+    catalog.finish();
+}
+
+fn write_and_collect_destinations<'a>(
+    ctx: &mut PdfContext,
+) -> Vec<(Str<'a>, Ref, Ref, f32, f32)> {
+    let mut destinations = vec![];
+
+    let elements = ctx.document.introspector.query(&HeadingElem::elem().select());
+    for elem in elements.iter() {
+        let heading = elem.to_packed::<HeadingElem>().unwrap();
+        if_chain!(
+            if let Some(label) = heading.label();
+            if let Some(first) = ctx.document.introspector.query_first(&Selector::And(eco_vec![
+                HeadingElem::elem().select(),
+                Selector::Label(label)
+            ]));
+            if heading == first.to_packed::<HeadingElem>().unwrap();
+            if let Some(loc) = heading.location();
```

Comment:
Queryable elements will always have a location, so we might as well unwrap here.